File: DFS/22.generate-parentheses.go
package leetcode

import "strings"

// @leet start
func generateParenthesis(n int) []string {
	var (
		ret, path []string
		dfs       func(l, r int)
	)
	dfs = func(l, r int) {
		if l == r && l == n {
			ret = append(ret, strings.Join(path, ""))
			return
		}
		if l < n {
			path = append(path, "(")
			dfs(l+1, r)
			path = path[:len(path)-1]
		}
		if r < l {
			path = append(path, ")")
			dfs(l, r+1)
			path = path[:len(path)-1]
		}
	}
	dfs(0, 0)
	return ret
}

// @leet end

File: DFS/22.generate-parentheses.py
from typing import List


# @leet start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def dfs(l, r, path):
            if l == r == n:
                res.append("".join(path))
                return
            if l < n:
                path.append("(")
                dfs(l + 1, r, path)
                path.pop()
            if l > r:
                path.append(")")
                dfs(l, r + 1, path)
                path.pop()

        dfs(0, 0, [])
        return res


# @leet end

